File: backend/app/schemas/image_dimensions.py
from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Annotated, Literal, TypeAlias, cast

from pydantic import AfterValidator

SEEDREAM_IMAGE_PRESET_SIZES = ("1K", "1.5K", "2K")
SEEDREAM_IMAGE_ASPECT_RATIOS = ("1:1", "4:3", "3:4", "16:9", "9:16")

SEEDREAM_MIN_IMAGE_PIXELS = 921_600
SEEDREAM_MAX_IMAGE_PIXELS = 4_624_220
SEEDREAM_MAX_ASPECT_RATIO = 16

SeedreamImagePresetSize: TypeAlias = Literal["1K", "1.5K", "2K"]
SeedreamImageAspectRatio: TypeAlias = Literal[
    "1:1",
    "4:3",
    "3:4",
    "16:9",
    "9:16",
]

_CUSTOM_IMAGE_SIZE_PATTERN = re.compile(r"^([0-9]+)x([0-9]+)$")


class SeedreamImageDimensionError(ValueError):
    def __init__(self, code: str, message: str) -> None:
        self.code = code
        super().__init__(message)


@dataclass(frozen=True, slots=True)
class SeedreamImageDimensions:
    width: int
    height: int
    pixels: int
    size: str


def is_seedream_image_preset_size(
    value: object,
) -> bool:
    return isinstance(value, str) and value in SEEDREAM_IMAGE_PRESET_SIZES


def parse_seedream_custom_image_size(
    value: object,
) -> SeedreamImageDimensions:
    if not isinstance(value, str):
        raise SeedreamImageDimensionError(
            "invalid_format",
            "image size must use WIDTHxHEIGHT decimal notation",
        )
    match = _CUSTOM_IMAGE_SIZE_PATTERN.fullmatch(value)
    if match is None:
        raise SeedreamImageDimensionError(
            "invalid_format",
            "image size must use WIDTHxHEIGHT decimal notation",
        )

    width = int(match.group(1))
    height = int(match.group(2))
    if width == 0 or height == 0:
        raise SeedreamImageDimensionError(
            "non_positive",
            "image width and height must be positive integers",
        )

    pixels = width * height
    if not SEEDREAM_MIN_IMAGE_PIXELS <= pixels <= SEEDREAM_MAX_IMAGE_PIXELS:
        raise SeedreamImageDimensionError(
            "pixel_count_out_of_range",
            "image pixel count must be between "
            f"{SEEDREAM_MIN_IMAGE_PIXELS} and {SEEDREAM_MAX_IMAGE_PIXELS}",
        )
    if (
        width > SEEDREAM_MAX_ASPECT_RATIO * height
        or height > SEEDREAM_MAX_ASPECT_RATIO * width
    ):
        raise SeedreamImageDimensionError(
            "aspect_ratio_out_of_range",
            "image aspect ratio must be between 1:16 and 16:1",
        )

    return SeedreamImageDimensions(
        width=width,
        height=height,
        pixels=pixels,
        size=f"{width}x{height}",
    )


def normalize_seedream_custom_image_size(value: object) -> str:
    return parse_seedream_custom_image_size(value).size


def normalize_seedream_image_size(
    value: object,
) -> SeedreamImageSize:
    if is_seedream_image_preset_size(value):
        return cast(SeedreamImagePresetSize, value)
    return normalize_seedream_custom_image_size(value)


SeedreamCustomImageSize: TypeAlias = Annotated[
    str,
    AfterValidator(normalize_seedream_custom_image_size),
]
SeedreamImageSize: TypeAlias = SeedreamImagePresetSize | SeedreamCustomImageSize


def _dimensions(value: str) -> SeedreamImageDimensions:
    return parse_seedream_custom_image_size(value)


SEEDREAM_COMMON_IMAGE_DIMENSIONS: dict[
    SeedreamImagePresetSize,
    dict[SeedreamImageAspectRatio, SeedreamImageDimensions],
] = {
    "1K": {
        "1:1": _dimensions("1024x1024"),
        "4:3": _dimensions("1152x864"),
        "3:4": _dimensions("864x1152"),
        "16:9": _dimensions("1424x800"),
        "9:16": _dimensions("800x1424"),
    },
    "1.5K": {
        "1:1": _dimensions("1536x1536"),
        "4:3": _dimensions("1792x1344"),
        "3:4": _dimensions("1344x1792"),
        "16:9": _dimensions("2048x1152"),
        "9:16": _dimensions("1152x2048"),
    },
    "2K": {
        "1:1": _dimensions("2048x2048"),
        "4:3": _dimensions("2368x1776"),
        "3:4": _dimensions("1776x2368"),
        "16:9": _dimensions("2816x1584"),
        "9:16": _dimensions("1584x2816"),
    },
}

File: frontend/lib/aigc/image-dimensions.ts
export const SEEDREAM_IMAGE_PRESET_SIZES = ["1K", "1.5K", "2K"] as const;
export const SEEDREAM_IMAGE_ASPECT_RATIOS = [
  "1:1",
  "4:3",
  "3:4",
  "16:9",
  "9:16"
] as const;

export const SEEDREAM_MIN_IMAGE_PIXELS = 921_600;
export const SEEDREAM_MAX_IMAGE_PIXELS = 4_624_220;
export const SEEDREAM_MAX_ASPECT_RATIO = 16;

export type SeedreamImagePresetSize =
  (typeof SEEDREAM_IMAGE_PRESET_SIZES)[number];
export type SeedreamImageAspectRatio =
  (typeof SEEDREAM_IMAGE_ASPECT_RATIOS)[number];

declare const customImageSizeBrand: unique symbol;
export type SeedreamCustomImageSize = string & {
  readonly [customImageSizeBrand]: true;
};
export type SeedreamImageSize =
  | SeedreamImagePresetSize
  | SeedreamCustomImageSize;

export type SeedreamImageDimensionErrorCode =
  | "invalid_format"
  | "non_positive"
  | "pixel_count_out_of_range"
  | "aspect_ratio_out_of_range";

export class SeedreamImageDimensionError extends Error {
  constructor(
    readonly code: SeedreamImageDimensionErrorCode,
    message: string
  ) {
    super(message);
    this.name = "SeedreamImageDimensionError";
  }
}

export interface SeedreamImageDimensions {
  readonly width: number;
  readonly height: number;
  readonly pixels: number;
  readonly size: SeedreamCustomImageSize;
}

const CUSTOM_IMAGE_SIZE_PATTERN = /^([0-9]+)x([0-9]+)$/;
const MIN_IMAGE_PIXELS_BIGINT = BigInt(SEEDREAM_MIN_IMAGE_PIXELS);
const MAX_IMAGE_PIXELS_BIGINT = BigInt(SEEDREAM_MAX_IMAGE_PIXELS);
const MAX_ASPECT_RATIO_BIGINT = BigInt(SEEDREAM_MAX_ASPECT_RATIO);

export function isSeedreamImagePresetSize(
  value: unknown
): value is SeedreamImagePresetSize {
  return (
    typeof value === "string" &&
    SEEDREAM_IMAGE_PRESET_SIZES.some((preset) => preset === value)
  );
}

export function parseSeedreamCustomImageSize(
  value: unknown
): SeedreamImageDimensions {
  if (typeof value !== "string") {
    throw new SeedreamImageDimensionError(
      "invalid_format",
      "Image size must use WIDTHxHEIGHT decimal notation"
    );
  }
  const match = CUSTOM_IMAGE_SIZE_PATTERN.exec(value);
  if (!match) {
    throw new SeedreamImageDimensionError(
      "invalid_format",
      "Image size must use WIDTHxHEIGHT decimal notation"
    );
  }

  const width = BigInt(match[1]);
  const height = BigInt(match[2]);
  if (width === BigInt(0) || height === BigInt(0)) {
    throw new SeedreamImageDimensionError(
      "non_positive",
      "Image width and height must be positive integers"
    );
  }

  const pixels = width * height;
  if (
    pixels < MIN_IMAGE_PIXELS_BIGINT ||
    pixels > MAX_IMAGE_PIXELS_BIGINT
  ) {
    throw new SeedreamImageDimensionError(
      "pixel_count_out_of_range",
      `Image pixel count must be between ${SEEDREAM_MIN_IMAGE_PIXELS} and ${SEEDREAM_MAX_IMAGE_PIXELS}`
    );
  }
  if (
    width > MAX_ASPECT_RATIO_BIGINT * height ||
    height > MAX_ASPECT_RATIO_BIGINT * width
  ) {
    throw new SeedreamImageDimensionError(
      "aspect_ratio_out_of_range",
      "Image aspect ratio must be between 1:16 and 16:1"
    );
  }

  const normalized = `${width}x${height}` as SeedreamCustomImageSize;
  return {
    width: Number(width),
    height: Number(height),
    pixels: Number(pixels),
    size: normalized
  };
}

export function normalizeSeedreamCustomImageSize(
  value: unknown
): SeedreamCustomImageSize {
  return parseSeedreamCustomImageSize(value).size;
}

export function isSeedreamCustomImageSize(
  value: unknown
): value is SeedreamCustomImageSize {
  try {
    parseSeedreamCustomImageSize(value);
    return true;
  } catch {
    return false;
  }
}

export function normalizeSeedreamImageSize(
  value: unknown
): SeedreamImageSize {
  if (isSeedreamImagePresetSize(value)) return value;
  return normalizeSeedreamCustomImageSize(value);
}

function dimensions(value: string): SeedreamImageDimensions {
  return parseSeedreamCustomImageSize(value);
}

export const SEEDREAM_COMMON_IMAGE_DIMENSIONS = {
  "1K": {
    "1:1": dimensions("1024x1024"),
    "4:3": dimensions("1152x864"),
    "3:4": dimensions("864x1152"),
    "16:9": dimensions("1424x800"),
    "9:16": dimensions("800x1424")
  },
  "1.5K": {
    "1:1": dimensions("1536x1536"),
    "4:3": dimensions("1792x1344"),
    "3:4": dimensions("1344x1792"),
    "16:9": dimensions("2048x1152"),
    "9:16": dimensions("1152x2048")
  },
  "2K": {
    "1:1": dimensions("2048x2048"),
    "4:3": dimensions("2368x1776"),
    "3:4": dimensions("1776x2368"),
    "16:9": dimensions("2816x1584"),
    "9:16": dimensions("1584x2816")
  }
} as const satisfies Readonly<
  Record<
    SeedreamImagePresetSize,
    Readonly<Record<SeedreamImageAspectRatio, SeedreamImageDimensions>>
  >
>;
